core/src/main/scala/org/apache/spark/executor/Executor.scala

```diff
@@ -218,6 +218,7 @@ private[spark] class Executor(
           threwException = false
           res
         } finally {
+          env.blockManager.releaseAllPinsForTask(taskId)
```

Comment:
See comment below; while I've gone ahead and added the machinery for this, we can't turn it on for all tests right now.